src/char/pincode.c

```diff
@@ -97,35 +137,93 @@ int pincode_compare(int fd, struct char_session_data* sd, char* pin) {
 	}
 }
 
-void pincode_change(int fd, struct char_session_data* sd) {
+void pincode_change(int fd, struct char_session_data* sd)
+{
 	char oldpin[5] = "\0\0\0\0", newpin[5] = "\0\0\0\0";
 
 	nullpo_retv(sd);
+
 	if (strlen(sd->pincode) != 4)
 		return;
+
 	safestrncpy(oldpin, RFIFOP(fd,6), sizeof(oldpin));
-	pincode->decrypt(sd->pincode_seed,oldpin);
-	if( !pincode->compare( fd, sd, oldpin ) )
+	pincode->decrypt(sd->pincode_seed, oldpin);
+
+	if (!pincode->compare(fd, sd, oldpin)) {
+		pincode->editstate(fd, sd, PINCODE_EDIT_FAILED);
+		pincode->loginstate(fd, sd, PINCODE_LOGIN_ASK);
 		return;
+	}
 
-	safestrncpy(newpin, RFIFOP(fd,10), sizeof(newpin));
-	pincode->decrypt(sd->pincode_seed,newpin);
-	pincode->update( sd->account_id, newpin );
-	safestrncpy(sd->pincode, newpin, sizeof(sd->pincode));
-	pincode->sendstate( fd, sd, PINCODE_ASK );
+	safestrncpy(newpin, RFIFOP(fd, 10), sizeof(newpin));
+	pincode->decrypt(sd->pincode_seed, newpin);
+
+	if (pincode->check_blacklist && pincode->is_blacklisted(newpin)) {
+		pincode->editstate(fd, sd, PINCODE_EDIT_RESTRICT_PW);
+	} else {
+		pincode->update(sd->account_id, newpin);
+		safestrncpy(sd->pincode, newpin, sizeof(sd->pincode));
+		pincode->editstate(fd, sd, PINCODE_EDIT_SUCCESS);
+	}
+	pincode->loginstate(fd, sd, PINCODE_LOGIN_ASK);
 }
 
-void pincode_setnew(int fd, struct char_session_data* sd) {
+void pincode_setnew(int fd, struct char_session_data* sd)
+{
 	char newpin[5] = "\0\0\0\0";
 
 	nullpo_retv(sd);
+
 	if (strlen(sd->pincode) == 4)
 		return;
+
 	safestrncpy(newpin, RFIFOP(fd,6), sizeof(newpin));
-	pincode->decrypt(sd->pincode_seed,newpin);
-	pincode->update( sd->account_id, newpin );
+	pincode->decrypt(sd->pincode_seed, newpin);
+
+	if (pincode->check_blacklist && pincode->is_blacklisted(newpin)) {
+		pincode->makestate(fd, sd, PINCODE_MAKE_RESTRICT_PW);
+		return;
+	}
+
+	pincode->update(sd->account_id, newpin);
 	safestrncpy(sd->pincode, newpin, sizeof(sd->pincode));
-	pincode->sendstate( fd, sd, PINCODE_ASK );
+	pincode->makestate(fd, sd, PINCODE_MAKE_SUCCESS);
+}
+
+/**
+* Send state of making new pincode
+*
+* @param[in] fd
+* @param[in, out] sd Session Data
+* @param[in] state Pincode Edit State
+*/
+void pincode_makestate(int fd, struct char_session_data *sd, uint16 state)
```

Comment:
probably better use created enums here?
replace ``uint16 state`` to ``enum pincode_make_response state``

Reply:
Fixed.